# syntax-transform/3.1/0003-singleton-class-superclass.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, the superclass of a singleton class is a subclass of `Class`
# rather than the superclass of a singleton class being the class of the
# singleton class's object. For example, this code snippet prints `nil`,
# `false`, `true` and `false` in Ruby 1.8 but `true`, `true`, `nil` and `false`
# in Ruby versions later than 1.8.
#
#   class A; end
#   class << A.new
#     p self < A
#     p self.superclass == A
#     p self < Class
#     p self.superclass == Class
#   end

--- a/class.c
+++ b/class.c
@@ -2026,7 +2026,7 @@ rb_special_singleton_class(VALUE obj)
  *       consistency of the metaclass hierarchy.
  */
 static VALUE
-singleton_class_of(VALUE obj)
+mkxp_singleton_class_of(VALUE obj, bool legacy)
 {
     VALUE klass;
 
@@ -2055,7 +2055,17 @@ singleton_class_of(VALUE obj)
     if (!(FL_TEST(klass, FL_SINGLETON) &&
           rb_attr_get(klass, id_attached) == obj)) {
 	rb_serial_t serial = RCLASS_SERIAL(klass);
-	klass = rb_make_metaclass(obj, klass);
+        if (legacy) {
+            VALUE metaclass = rb_class_boot(klass);
+            FL_SET(metaclass, FL_SINGLETON);
+            rb_singleton_class_attached(metaclass, obj);
+            SET_METACLASS_OF(obj, metaclass);
+            SET_METACLASS_OF(metaclass, metaclass);
+            RCLASS_SET_SUPER(metaclass, RBASIC(rb_class_real(klass))->klass);
+            klass = metaclass;
+        } else {
+	    klass = rb_make_metaclass(obj, klass);
+        }
 	RCLASS_SERIAL(klass) = serial;
     }
 
@@ -2064,6 +2074,12 @@ singleton_class_of(VALUE obj)
     return klass;
 }
 
+static VALUE
+singleton_class_of(VALUE obj)
+{
+    return mkxp_singleton_class_of(obj, FALSE);
+}
+
 void
 rb_freeze_singleton_class(VALUE x)
 {
@@ -2099,9 +2115,9 @@ rb_singleton_class_get(VALUE obj)
 }
 
 VALUE
-rb_singleton_class(VALUE obj)
+mkxp_rb_singleton_class(VALUE obj, bool legacy)
 {
-    VALUE klass = singleton_class_of(obj);
+    VALUE klass = mkxp_singleton_class_of(obj, legacy);
 
     /* ensures an exposed class belongs to its own eigenclass */
     if (RB_TYPE_P(obj, T_CLASS)) (void)ENSURE_EIGENCLASS(klass);
@@ -2109,6 +2125,12 @@ rb_singleton_class(VALUE obj)
     return klass;
 }
 
+VALUE
+rb_singleton_class(VALUE obj)
+{
+    return mkxp_rb_singleton_class(obj, FALSE);
+}
+
 /*!
  * \}
  */
--- a/insns.def
+++ b/insns.def
@@ -709,7 +709,7 @@ defineclass
 (VALUE cbase, VALUE super)
 (VALUE val)
 {
-    VALUE klass = vm_find_or_create_class_by_id(id, flags, cbase, super);
+    VALUE klass = vm_find_or_create_class_by_id(id, flags, cbase, super, class_iseq);
 
     rb_iseq_check(class_iseq);
 
--- a/vm_insnhelper.c
+++ b/vm_insnhelper.c
@@ -4612,7 +4612,7 @@ static VALUE
 vm_find_or_create_class_by_id(ID id,
 			      rb_num_t flags,
 			      VALUE cbase,
-			      VALUE super)
+			      VALUE super, const rb_iseq_t *iseq)
 {
     rb_vm_defineclass_type_t type = VM_DEFINECLASS_TYPE(flags);
 
@@ -4623,7 +4623,8 @@ vm_find_or_create_class_by_id(ID id,
 
       case VM_DEFINECLASS_TYPE_SINGLETON_CLASS:
 	/* classdef returns class scope value */
-	return rb_singleton_class(cbase);
+	VALUE mkxp_rb_singleton_class(VALUE, bool);
+	return mkxp_rb_singleton_class(cbase, mkxp_is_syntax_transform_active(iseq, 1, 8));
 
       case VM_DEFINECLASS_TYPE_MODULE:
 	/* classdef returns class scope value */
